docs/guides/javascript.rst

```diff
@@ -12,88 +12,159 @@ The 1.8 version is still functional and is :ref:`described below<1.8-js>`.
-------------------------------------------------
+Once a module is defined, telling Elgg to load it in the current page is easy:
 
-You can define a module by creating a view or registering a URL.
+.. code-block:: php
+
+    <?php
+    elgg_require_js("myplugin/say_hello");
+
+On the client-side, this will asynchronously load the module, load any dependencies, and
+execute it.
+
+This can also be done from JavaScript by calling ``require(["myplugin/say_hello"])``,
+but you should not need to.
+
+Defining the Module
+===================
 
-Defining modules as a view
-^^^^^^^^^^^^^^^^^^^^^^^^^^
+A module is essentially a list of dependencies and a function which is executed, maybe
+returning a value to other modules. It can perform operations on the page, return an API
+or value for consumption by other modules, or both.
 
-Modules defined by creating views are immediately available for use and require no registration.
-To register a module named ``my/module``, create the view ``views/default/js/my/module.js``.
+Here we define a basic module ``myplugin/say_hello`` for altering the page:
+
+.. code-block:: javascript
+
+    // in views/default/js/myplugin/say_hello.js
+
+    define(function(require) {
+        var elgg = require("elgg");
+        var $ = require("jquery");
+
+        $('body').append(elgg.echo('hello_world'));
+    });
+
+.. note::
+
+    RequireJS uses a bit of magic to recognize the ``elgg`` and ``jquery`` modules as
+    dependencies, even before executing the function!
+
+The module is implemented as the view ``js/myplugin/say_hello.js`` but, since it's a static JS file, it's
+available client-side without registering it.
 
 .. warning: The extension must be ``.js``.
 
-A basic module could look like this:
+Making modules dependent on other modules
+-----------------------------------------
+
+Below we refactor a bit so that the module depends on a new ``myplugin/hello`` module to provide
+the greeting:
 
 .. code-block:: javascript
 
-	define(function(require) {
-		var elgg = require("elgg");
-		var $ = require("jquery");
+    // in views/default/js/myplugin/hello.js
 
-		return function() {
-			// Some logic in here
-		};
-	});
+    define(function(require) {
+        var elgg = require("elgg");
 
-Define your module via a URL
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        return elgg.echo('hello_world');
+    });
 
-You can define an existing AMD module using ``elgg_define_js()``. Traditional (browser-globals)
-JavaScript files can also be defined as AMD modules if you shim them by setting ``exports`` and
-optionally ``deps``.
+.. code-block:: javascript
+
+    // in views/default/js/myplugin/say_hello.js
+
+    define(function(require) {
+        var $ = require("jquery");
+        var hello = require("myplugin/hello");
+
+        $('body').append(hello);
+    });
 
-.. warning:: Calls to ``elgg_define_js()`` must be in an ``init, system`` event handler.
+
+Defining a module with a PHP view
```

Comment:
This might be another one to drop -- what is the use-case for such a feature? Won't think just cause more confusion that necessary if people are confused why `elgg_get_logged_in_user_guid()` is not working as they expect when they use it in a JS file?